esrally/metrics.py

```diff
@@ -1673,6 +1675,11 @@ def total_transform_metric(self, metric_name):
     def error_rate(self, task_name, operation_type):
         return self.store.get_error_rate(task=task_name, operation_type=operation_type, sample_type=SampleType.Normal)
 
+    def duration(self, task_name):
+        values = self.store.get_raw("service_time", task_name, mapper=lambda doc: doc["relative-time"])
```

Comment:
This can potentially return millions of documents (e.g. in a benchmark with thousands of clients) and we're only interested in one value. I wonder whether this would justify a specialized query method in the metrics store as it would be much cheaper only to retrieve the maximum value directly?

I also want to point out that `relative-time` is determined *before* a request is issued:

https://github.com/elastic/rally/blob/62595bde7319b77895bc8c5b726c9cd298daef30/esrally/driver/driver.py#L1060-L1062

So strictly speaking `duration` does not account for the service time of the last request. If this is a long-running request (e.g. in frozen tier benchmarks) the duration measurement could be several minutes off. This could be specifically misleading if there is only one measurement iteration as `duration` would report something in order of several milliseconds when the actual duration would more be like several minutes.

So if we settle on a special query method, I propose that the API allows us to retrieve one sample ordered by a criterion. E.g. we could amend the current `get_one` method in the metrics store:

https://github.com/elastic/rally/blob/62595bde7319b77895bc8c5b726c9cd298daef30/esrally/metrics.py#L619

to allow sorting:

```python
def get_one(self, name, sample_type=None, node_name=None, task=None, sort_key=None, sort_reverse=False):
```

_Note: I've borrowed the names of the new parameters `sort_key` an `sort_reverse` from Python's [`sorted`](https://docs.python.org/3/library/functions.html#sorted) function for an idiomatic API._

What do you think?

Reply:
Sounds good, I will use that, thank you for the suggestion!

Reply:
@danielmitterdorfer interestingly the current implementation for get_one uses get and the just returns values[0] so it would still return all the values and then pick the first one locally.
https://github.com/elastic/rally/blob/62595bde7319b77895bc8c5b726c9cd298daef30/esrally/metrics.py#L633
 Was your idea to modify the search query so that it would only return one value?

Reply:
Yes, we should implement `get_one` in both subclasses (`InMemoryMetricsStore` and `EsMetricsStore`) and remove the generic implementation in the base class.